File: packages/go-sdk/forms_test.go
package kibamail

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestFormsCreate(t *testing.T) {
	client := setupTest()

	t.Run("should create a form with basic information", func(t *testing.T) {
		result, err := client.Forms.Create(&CreateFormRequest{
			Name:        "Newsletter Signup",
			Description: "Subscribe to our weekly newsletter",
		})

		if err != nil {
			t.Skipf("Skipping test due to API error: %v", err)
			return
		}
		assert.NotNil(t, result)
		assert.NotEmpty(t, result.ID)
	})

	t.Run("should create a form with fields", func(t *testing.T) {
		result, err := client.Forms.Create(&CreateFormRequest{
			Name:        "Contact Form",
			Description: "General contact form",
			Fields: []interface{}{
				map[string]interface{}{
					"name":     "email",
					"type":     "email",
					"required": true,
				},
				map[string]interface{}{
					"name":     "firstName",
					"type":     "text",
					"required": false,
				},
			},
		})

		if err != nil {
			t.Skipf("Skipping test due to API error: %v", err)
			return
		}
		assert.NotNil(t, result)
		assert.NotEmpty(t, result.ID)
	})
}

func TestFormsList(t *testing.T) {
	client := setupTest()

	t.Run("should list forms with default pagination", func(t *testing.T) {
		result, err := client.Forms.List(nil)

		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.NotNil(t, result.Data)
	})

	t.Run("should list forms with custom limit", func(t *testing.T) {
		limit := 50
		result, err := client.Forms.List(&ListOptions{Limit: &limit})

		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.NotNil(t, result.Data)
	})
}

func TestFormsGet(t *testing.T) {
	client := setupTest()

	t.Run("should retrieve a form by ID", func(t *testing.T) {
		formId := "form_test_12345"
		result, err := client.Forms.Get(formId)

		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.NotEmpty(t, result.ID)
	})
}

func TestFormsUpdate(t *testing.T) {
	client := setupTest()

	t.Run("should update a form's basic information", func(t *testing.T) {
		formId := "form_test_12345"
		result, err := client.Forms.Update(formId, &UpdateFormRequest{
			Name:        "Updated Form",
			Description: "Updated description",
		})

		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.NotEmpty(t, result.ID)
	})

	t.Run("should update a form's fields", func(t *testing.T) {
		formId := "form_test_12345"
		result, err := client.Forms.Update(formId, &UpdateFormRequest{
			Fields: []interface{}{
				map[string]interface{}{
					"name":     "email",
					"type":     "email",
					"required": true,
				},
			},
		})

		assert.NoError(t, err)
		assert.NotNil(t, result)
		assert.NotEmpty(t, result.ID)
	})
}

func TestFormsDelete(t *testing.T) {
	client := setupTest()

	t.Run("should delete a form by ID", func(t *testing.T) {
		formId := "form_test_12345"
		err := client.Forms.Delete(formId)

		assert.NoError(t, err)
	})
}

File: sdks/nodejs/tests/forms.test.ts
import { describe, it, expect, beforeEach } from "vitest";
import { Kibamail } from "../kibamail";
import { MOCK_API_URL, MOCK_API_KEY } from "./setup";

describe("Forms Resource", () => {
  let kibamail: Kibamail;

  beforeEach(() => {
    kibamail = new Kibamail(MOCK_API_KEY, {
      baseURL: MOCK_API_URL,
    });
  });

  describe("create", () => {
    it("should create a form with name and fieldMapping", async () => {
      const result = await kibamail.forms.create({
        name: "Newsletter Signup",
        fieldMapping: {
          email: {
            contactPropertyId: "email",
            contactPropertyType: "standard",
            fieldType: "string",
          },
        },
      });

      expect(result.data).toBeDefined();
      expect(result.error).toBeUndefined();
      expect(result.response.status).toBe(201);
    });
  });

  describe("list", () => {
    it("should list all forms", async () => {
      const result = await kibamail.forms.list();

      expect(result.data).toBeDefined();
      expect(result.error).toBeUndefined();
      expect(result.response.status).toBe(200);
    });

    it("should list forms with pagination", async () => {
      const result = await kibamail.forms.list({
        limit: 20,
        after: "form_abc123",
      });

      expect(result.data).toBeDefined();
      expect(result.error).toBeUndefined();
    });
  });

  describe("get", () => {
    it("should retrieve a form by ID", async () => {
      const formId = "form_test_12345";
      const result = await kibamail.forms.get(formId);

      expect(result.data).toBeDefined();
      expect(result.error).toBeUndefined();
      expect(result.response.status).toBe(200);
    });
  });

  describe("update", () => {
    it("should update a form's name", async () => {
      const formId = "form_test_12345";
      const result = await kibamail.forms.update(formId, {
        name: "Updated Form Name",
      });

      expect(result.data).toBeDefined();
      expect(result.error).toBeUndefined();
    });
  });

  describe("delete", () => {
    it("should delete a form by ID", async () => {
      const formId = "form_test_12345";
      const result = await kibamail.forms.delete(formId);

      expect(result.data).toBeDefined();
      expect(result.error).toBeUndefined();
      expect(result.response.status).toBe(200);
    });
  });

  describe("publish", () => {
    it("should publish a form", async () => {
      const formId = "form_test_12345";
      const result = await kibamail.forms.publish(formId);

      expect(result.data).toBeDefined();
      expect(result.error).toBeUndefined();
      expect(result.response.status).toBe(200);
    });
  });

  describe("listVersions", () => {
    it("should list all versions of a form", async () => {
      const formId = "form_test_12345";
      const result = await kibamail.forms.listVersions(formId);

      expect(result.data).toBeDefined();
      expect(result.error).toBeUndefined();
      expect(result.response.status).toBe(200);
    });
  });
});
